chain/starknet/src/data_source.rs

```diff
@@ -144,6 +175,26 @@ impl blockchain::DataSource<Chain> for DataSource {
     }
 }
 
+impl DataSource {
+    /// Returns event trigger if an event.key matches the handler.key and optionally
+    /// if event.fromAddr matches the source address. Note this only supports the default
+    /// starknet behavior of one key per event.
+    fn handler_for_event(&self, event: &codec::Event) -> Option<MappingEventHandler> {
+        return self
+            .mapping
+            .event_handlers
+            .iter()
+            .find(|handler| match event.keys.first() {
+                Some(key) => match &self.source.address {
+                    Some(address) => address == &event.from_addr && key == &handler.key(),
+                    None => key == &handler.key(),
+                },
+                None => false,
+            })
```

Comment:
Here it's repeatedly checking whether `event` has at least one key. It should be done before calling `find` instead.